src/neuroglancer/annotation/frontend_source.ts

```diff
@@ -409,6 +443,120 @@ export class MultiscaleAnnotationSource extends SharedObject implements
     }
   }
 
+  activeAnnotations(state: AnnotationLayerState, sortByPosition: Position): [
+    length: number,
+    indexToId: (index: number) => Annotation,
+    idToIndex: (id: string) => number | undefined,
+  ] {
+    const {segmentFilteredSources, spatiallyIndexedSources, rank, properties, relationships} = this;
+    const {relationshipStates} = state.displayState;
+    let hasVisibleSegments = false;
+
+    let currentLength = 0;
+    const listOffsets: number[] = [];  // TODO, maybe change to index start
+    const serializedAnnotationsList: SerializedAnnotations[] = [];
+    const deserializedAnnotationsList: Annotation[][] = [];
+    const idToIndexMap = new Map<string, number>();
+
+    for (let i = 0; i < relationships.length; i++) {
+      const relationship = relationships[i];
+      const state = relationshipStates.get(relationship)
+      if (state) {
+        const {showMatches: {value: showMatches}, segmentationState: {value: segmentationState}} =
+            state;
+        if (!showMatches || !segmentationState) continue;
+        const chunks = segmentFilteredSources[i].chunks;
+        forEachVisibleSegment(segmentationState.segmentationGroupState.value, objectId => {
+          hasVisibleSegments = true;
+          const key = getObjectKey(objectId);
+          const chunk = chunks.get(key);
+          if (chunk !== undefined && chunk.state === ChunkState.GPU_MEMORY) {
+            const {data} = chunk;
+            if (data === undefined) return;
+            const {serializedAnnotations} = data;
+            const length = serializedAnnotations.typeToIds.reduce(
+                (sum, idsOfType) => sum + idsOfType.length, 0);
+            listOffsets.push(currentLength);
+            currentLength += length;
+          }
+        });
+      }
+    }
+    if (!hasVisibleSegments) {
+      for (const source of spatiallyIndexedSources) {
+        const {rank} = sortByPosition.coordinateSpace.value;
+        const sortChunk = new Uint32Array(rank);
+        const {chunkDataSize} = source.spec;
+        for (let i = 0; i < rank; i++) {
+          sortChunk[i] = Math.floor(sortByPosition.value[i] / chunkDataSize[i]);
```

Comment:
This is the basic chunk calculation that doesn't match up with the loaded chunks. The idea is to start with the highest level chunks and go out till we have a chunk that is loaded. Eventually it would be nice to support neighboring chunks particularly if the global is near a chunk boundary.

Reply:
For `sortByPosition` you are just using the `globalPosition`, which is not necessarily in the same coordinate space.  Instead you need to transform the relevant globalPosition and localPosition to the coordinate space of the annotation layer, as done here:

https://github.com/google/neuroglancer/blob/c466b24a07cfba28d6766a9a5733fd4182ea25d2/src/neuroglancer/ui/annotations.ts#L852

Then each scale level also has a coordinate transform.

Reply:
I converted `getMousePositionInAnnotationCoordinates` to `getPositionInAnnotationCoordinates` so I could pass in any Float32Array, call that with
```ts
const point = getPositionInAnnotationCoordinates(
  globalPosition.value,
  state,
);
```

I then copied the logic in `forEachPossibleChunk`:
https://github.com/google/neuroglancer/blob/c466b24a07cfba28d6766a9a5733fd4182ea25d2/src/neuroglancer/annotation/frontend_source.ts#L567
To make `positionToMultiscaleChunks(position: Float32Array)`
https://gist.github.com/chrisj/328828d13480627961166beb16cfcd18

This uses `source.multiscaleToChunkTransform` which I assume is the coordinate transform that you mentioned per scale. The issue is that it only returns 0,0,0 for the largest scale. I'm not sure if I'm misunderstanding the code but what happens in the case of a single point is that `totalChunks = 1;` (unless on boundary), `remainder = 0`
`chunkIndex = 0;`, `size = 1`, `remainder % size = 0`, `remainder = 0 / 1` so you end up with tempChunk = [0,0,0] and that chunk usually only exists for the largest scale.

Reply:
Can you explain a bit more what the issue is?  It is indeed the case that at the largest scale, there is normally just a single chunk.

Reply:
I assume that MultiscaleAnnotationSource.forEachPossibleChunk is supposed run the callback across all the scales that contain the annotation.

When I follow the logic, it seems to be stuck at [0,0,0] across all scales so `chunks.get(tempChunk.join());` only returns a result at the largest scale.

I am looking for the smallest scales that contain my target point.

Reply:
I added this branch that shows the problem I'm experiencing
https://github.com/seung-lab/neuroglancer/tree/cj_multiscale_annotation_list_3

With this link http://127.0.0.1:8080/#!%7B%22dimensions%22:%7B%22x%22:%5B8e-9%2C%22m%22%5D%2C%22y%22:%5B8e-9%2C%22m%22%5D%2C%22z%22:%5B8e-9%2C%22m%22%5D%7D%2C%22position%22:%5B15610.5%2C22520.5%2C24095.5%5D%2C%22crossSectionScale%22:1%2C%22projectionOrientation%22:%5B-0.29930439591407776%2C-0.3578258454799652%2C0.8785958290100098%2C-0.10221008211374283%5D%2C%22projectionScale%22:12.370998330885378%2C%22layers%22:%5B%7B%22type%22:%22annotation%22%2C%22source%22:%22precomputed://gs://neuroglancer-janelia-flyem-hemibrain/v1.0/synapses%22%2C%22tab%22:%22annotations%22%2C%22shader%22:%22#uicontrol%20vec3%20preColor%20color%28default=%5C%22red%5C%22%29%5Cn#uicontrol%20vec3%20postColor%20color%28default=%5C%22blue%5C%22%29%5Cn#uicontrol%20float%20preConfidence%20slider%28min=0%2C%20max=1%2C%20default=0%29%5Cn#uicontrol%20float%20postConfidence%20slider%28min=0%2C%20max=1%2C%20default=0%29%5Cn%5Cnvoid%20main%28%29%20%7B%5Cn%20%20setColor%28defaultColor%28%29%29%3B%5Cn%20%20setEndpointMarkerColor%28%5Cn%20%20%20%20vec4%28preColor%2C%200.5%29%2C%5Cn%20%20%20%20vec4%28postColor%2C%200.5%29%29%3B%5Cn%20%20setEndpointMarkerSize%282.0%2C%202.0%29%3B%5Cn%20%20setLineWidth%282.0%29%3B%5Cn%20%20if%20%28prop_pre_synaptic_confidence%28%29%3C%20preConfidence%20%7C%7C%5Cn%20%20%20%20%20%20prop_post_synaptic_confidence%28%29%3C%20postConfidence%29%20discard%3B%5Cn%7D%5Cn%22%2C%22filterBySegmentation%22:%5B%22post_synaptic_cell%22%2C%22pre_synaptic_cell%22%5D%2C%22name%22:%22synapse%22%7D%5D%2C%22showSlices%22:false%2C%22selectedLayer%22:%7B%22size%22:426%2C%22visible%22:true%2C%22layer%22:%22synapse%22%7D%2C%22layout%22:%223d%22%2C%22statistics%22:%7B%22size%22:248%7D%7D

In the console you will see output such as

```
addChunk 0,0,0
addChunk 28,31,35
addChunk 14,15,17
addChunk 0,1,1
addChunk 0,0,0
addChunk 1,0,1
addChunk 29,30,36
addChunk 28,32,36
addChunk 0,0,1
... (removed 20 lines)
addChunk 2,3,4
addChunk 7,7,9
addChunk 1,1,2
addChunk 1,1,1
5 no chunk for key 0,0,0
chunk exists 0,0,0 size 17209,15302,19814
```